src/array-with-default.js

```diff
@@ -37,41 +38,51 @@ export class ArrayWithDefault extends Array {
         return new Proxy(this, {
             get(target, property) {
 
-                const value = target[property];
+                let currValue = target[property];
 
-                // symbol properties excluded
+                // exit on symbol properties
                 if (typeof property !== "string") {
-                    return value;
+                    return currValue;
                 }
 
                 const index = Number(property);
 
                 /*
-                 * We only want to return the default when the index
-                 * is numeric (an array element), when the index is less
-                 * than the length of the array (indicating a missing element),
-                 * and when the value is undefined.
-                 * 
-                 * For an index beyond the length of the array, or a string or
-                 * symbol property, just return the regular value.
+                 * We only want to return the default when:
+                 * - index is numeric (an array element),
+                 * - currValue is undefined
+                 * - index is less than the length of the array, or out-of-range
+                 *
+                 * If default is a callback, assign its result to the missing element.
                  */
                 if (
-                    Number.isInteger(index) && 
-                    (options.outOfRange || index < target.length) &&
-                    value === undefined
+                    Number.isInteger(index) &&
+                    currValue === undefined &&
+                    (options.outOfRange || index < target.length)
                 ) {
-                    return options.default;
+                    let defaultValue = options.default;
+
+                    if (typeof defaultValue === "function") {
+                        // resolve the callback defaultValue
+                        defaultValue = defaultValue(index);
+                        return target[property] = defaultValue;
```

Comment:
We could also only assign to `target[property]` if the resolved `defaultValue` is an object. There are `this.length` implications either way. Thoughts?